raft/server.cc

```diff
@@ -454,6 +462,47 @@ future<> server_impl::wait_for_state_change(seastar::abort_source* as) {
     }
 }
 
+future<> server_impl::trigger_snapshot(seastar::abort_source& as) {
+    check_not_aborted();
+
+    if (_applied_idx <= _snapshot_desc_idx) {
+        logger.info(
+            "[{}] trigger_snapshot: last persisted snapshot descriptor index is up-to-date",
+            ", applied index: {}, persisted snapshot descriptor index: {}, last fsm log index: {}"
+            ", last fsm snapshot index: {}", _id, _applied_idx, _snapshot_desc_idx,
+            _fsm->log_last_idx(), _fsm->log_last_snapshot_idx());
+        co_return;
+    }
+
+    _new_server_requests.snapshot = true;
+    _events.signal();
+
+    // Wait for persisted snapshot index to catch up to this index.
+    auto awaited_idx = _applied_idx;
+
+    logger.info("[{}] snapshot request waiting for index {}", _id, awaited_idx);
+
+    try {
+        as.check();
+        auto sub = as.subscribe([this] () noexcept { _snapshot_desc_idx_changed.broadcast(); });
+        assert(sub); // due to `check()` above
+        co_await _snapshot_desc_idx_changed.when([this, &as, awaited_idx] {
+            return as.abort_requested() || awaited_idx <= _snapshot_desc_idx;
+        });
+        as.check();
+    } catch (abort_requested_exception&) {
+        throw request_aborted();
+    } catch (seastar::broken_condition_variable&) {
+        throw request_aborted();
+    }
+
+    logger.info(
```

Comment:
It's an external API and should be (very) rarely triggered, it won't cause any spam. From debugging problems in CI etc. I find that I'm usually missing logs, rarely I complain that there are too many of them. (Usually only if there's a hot loop somewhere)

We can demote it to debug in the future if we want to start calling it often.

(BTW. we're still pretending that this is a library that someone will use outside Scylla?)

Reply:
> It's an external API and should be (very) rarely triggered, it won't cause any spam. From debugging problems in CI etc. I find that I'm usually missing logs, rarely I complain that there are too many of them. (Usually only if there's a hot loop somewhere)

We can log where we call the function. We may have less info there, but we can log the fact that the snapshot was triggered and without raft trace indexes do not have much sense anyway.

> 
> We can demote it to debug in the future if we want to start calling it often.
> 
> (BTW. we're still pretending that this is a library that someone will use outside Scylla?)

Why not?

Reply:
> We can log where we call the function. We may have less info there, but we can log the fact that the snapshot was triggered and without raft trace indexes do not have much sense anyway.

Ok, I will demote these to DEBUG.

> Why not?

Because it's inside Scylla repo, for example.

Reply:
> Because it's inside Scylla repo, for example.

And licensed using AGPL